### PR TITLE
[JIT] apply() for ScriptModules

### PR DESCRIPTION
**Pull request by jamesr66a**
This can be use to initialize state that is not necessarily eligible for serialization/is implementation-specific. Concretely, I'm going to use this to pack the weight matrices for quantized Linear modules according to the FBGEMM APIs

### CODE REVIEW

**Review by apaszke — DISMISSED** [review]
`pack` and `unpack` definitely don't look like special methods, and can easily clash with user code. If we *really* want to go down this path, we should either:
1. make them private for internal use/testing this solution
2. create a full public API for repacking the weight format of a given `nn.Module` (e.g. according to the proposal I write last year)

**Comment by zdevito** [comment]
I think the approach is fine, but I think we should not put it in the core data structures and instead provide it off to the side. This is because real custom serialization will provide additional functionality and work differently that this proposal in a way that cannot be easily made backwards compatible. 

Concretely:

1. This diff should not change script::Module or torch.jit.ScriptModule.
2. We provide a new header that provides 2 C++  functions that just visit every module and call "_unpack" or "_pack". We bind these into python as well. I don't think there is a particular advantage in meta-programming the generation of a full recursive pack function. Consider adding the C++ equivalent of Python's Module.apply method for making each of these functions really short.
3. Where we want to use this pack/unpack convention, we simply call these functions before calling save.

Eventually:
We build a more complete and thought out api for custom serialization, and we switch uses of this convention over to that system.

**Comment by jamesr66a** [comment]
@zdevito I updated the PR to align with your proposal

File: test/test_jit.py [hunk 1]
```diff
@@ -5425,6 +5425,115 @@ def forward(self):
...
+            def _unpack(self):
+                self.buf.set_(torch.ones(3, 4, dtype=torch.double))
+
+            @torch.jit.script_method
+            def forward(self, x):
+                return self.submod(x + self.buf)
+
+        m = Mod()
+        torch.testing.assert_allclose(m(torch.zeros(3, 4)), torch.ones(3, 4) * 6)
+        torch._C._recursively_call_method(m, '_pack')
```
Comment (zdevito) [inline test/test_jit.py]: The python Module already has a `.apply` function:
`m.apply(lambda s: s._pack())`

File: torch/csrc/jit/script/module.h [hunk 2]
```diff
@@ -470,6 +470,20 @@ struct Module {
   bool optimize;
 };
 
+static inline void recursivelyCallMethod(std::shared_ptr<Module> mod, const std::string& method_name) {
```
Comment (zdevito) [inline torch/csrc/jit/script/module.h]: Can we just make this the equivalent of the Python modules apply function? C++ code would then be similar for unpack:

```
void script::Module::apply(std::function<void(Module&)> fn);
```

```
m.apply([](Module& m) {
  m.get_method("_unpack")({});
});
```

**Review by zdevito — COMMENTED** [review]
I think this can be made even simpler by just adding the equivalent of Python Module's `apply` to C++.

**Review by zdevito — APPROVED** [review]
